src/TransformOperationExecutor.ts

```diff
@@ -14,33 +15,52 @@ export class TransformOperationExecutor {
     // Private Properties
     // -------------------------------------------------------------------------
 
-    private transformedTypesMap = new Map<Object, {level: number, object: Object}>();
+    private recursionStack = new Set<Object>();
 
     // -------------------------------------------------------------------------
     // Constructor
     // -------------------------------------------------------------------------
 
     constructor(private transformationType: TransformationType,
-                private options: ClassTransformOptions) {
+        private options: ClassTransformOptions) {
     }
 
     // -------------------------------------------------------------------------
     // Public Methods
     // -------------------------------------------------------------------------
 
-    transform(source: Object|Object[]|any,
-              value: Object|Object[]|any,
-              targetType: Function,
-              arrayType: Function,
-              isMap: boolean,
-              level: number = 0) {
+    transform(source: Object | Object[] | any,
+        value: Object | Object[] | any,
+        targetType: Function | TypeMetadata,
```

Comment:
I am not happy about this overloading, this function is way bigger and messier than it should be, overloading I just add more to the confusion, but we can fix this after release.